app/abuser_utils.py

```diff
@@ -0,0 +1,197 @@
+import hmac
+import json
+import secrets
+from hashlib import sha256
+from typing import List, Dict, Optional
+
+from cryptography.exceptions import InvalidTag
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes as crypto_hashes
+from cryptography.hazmat.primitives.ciphers.aead import AESGCM
+from cryptography.hazmat.primitives.kdf.hkdf import HKDF
+
+from app import config
+from app.db import Session
+from app.log import LOG
+from app.models import User, Alias, Mailbox, AbuserData, AbuserLookup
+
+
+def _derive_encryption_key(master_key: bytes, user_id: int) -> bytes:
+    if user_id is None or not isinstance(user_id, int) or user_id <= 0:
+        raise ValueError("User ID must be a valid positive integer for key derivation.")
+
+    hkdf_info = (config.HKDF_INFO_TEMPLATE % str(user_id)).encode("utf-8")
+    hkdf = HKDF(
+        algorithm=crypto_hashes.SHA256(),
+        length=32,
+        salt=None,
+        info=hkdf_info,
+        backend=default_backend(),
+    )
+
+    return hkdf.derive(master_key)
+
+
+def check_if_abuser_email(new_address: str) -> Optional[AbuserLookup]:
+    """
+    Returns False, if the given address (after hashing) is found in abuser_lookup.
+    """
+    mac_key_bytes = config.MAC_KEY
+    check_hmac = hmac.new(
+        mac_key_bytes, new_address.encode("utf-8"), sha256
+    ).hexdigest()
+
+    return (
+        AbuserLookup.filter(AbuserLookup.hashed_address == check_hmac).limit(1).first()
+    )
+
+
+def archive_abusive_user(user: User) -> None:
+    """
+    Archive the given abusive user's data and update blocklist/lookup tables.
+    """
+    if not user.email:
+        raise ValueError(f"User ID {user.id} must have a primary email to be archived.")
+
+    try:
+        primary_email: str = user.email
+        aliases: List[Alias] = Alias.filter_by(user_id=user.id).all()
```

Comment:
User may have tons of alias. I'd recommend doing 
```suggestion
        aliases: List[Alias] = Alias.filter_by(user_id=user.id).enable_eagerloads(False).yield_per(500).all()
```
This way it will page the queries to the db behind the scenes